### ISSUE TITLE
A better async/await actions usage?

### ISSUE THREAD

**Issue by whinc** [mobxjs/mobx-state-tree#1415]
**_Question_**

* [x] I've checked documentation and searched for existing issues
* [ ] I tried the [spectrum channel](https://spectrum.chat/?t=dad48299-3dfc-4e10-b6da-9af1e39498a3)

I have read [Creating asynchronous actions](https://github.com/mobxjs/mobx-state-tree/blob/master/docs/async-actions.md). The main reason we can't use async/await is that modification operations to state tree in async/await function cann't be detech automatically. so, i think we can write like below to continue use async/await by explicitly wrap operations in a action.I am not sure if there are potential side effects? If it's ok, i hope MST can provider natively the `runInAction` interface to the model type
```javascript
  const Store = types.model({
        githubProjects: types.array(types.frozen),
        state: types.enumeration("State", ["pending", "done", "error"])
    })
    .actions(self => ({
        runInAction(action) {
          action.call(self)
        }
    }))
    .actions(self => ({
        async fetchProjects() {
            self.githubProjects = []
            self.state = "pending"
            try {
                const projects = await fetchGithubProjectsSomehow()
                self.runInAction(() => {
                    self.state = "done"
                    self.githubProjects = projects
                })
            } catch (error) {
                console.error("Failed to fetch projects", error)
                self.runInAction(() => {
                    self.state = "error"
                })
            }
        }
    }))
```

**Comment by impulse** [comment]
Not sure if I understood right but this runInAction wrapping is already done if you use flow.

**Comment by zmayor** [comment]
> Not sure if I understood right but this runInAction wrapping is already done if you use flow.

😨...`flow`  receive a generator function instead of an async function as it's parameter.

**Comment by beaulac** [comment]
Yes, but [as mentioned in the docs](https://github.com/mobxjs/mobx-state-tree/blob/master/docs/async-actions.md#what-about-async--await), it's as simple as replacing `async function x() {/*...*/}` with `flow(function* x() {/*...*/})`, and `await` with `yield`.

The generator version of this action in the docs is shorter and simpler than the one using async+`runInAction`, _and_ unlike async implementation it actually counts as **1** action (which it should!). This is an important part of why you shouldn't use the proposed async+`runInAction` pattern:

Imagine you have an `onAction` handler for your instance. If your code outside calls `fetchProjects` once, you would expect that `onAction` gets triggered once, but each  `runInAction` step triggers an additional `onAction` (see [this sandbox](https://codesandbox.io/s/peaceful-lederberg-8yuul)) which is not ideal / expected behavior. 

This also breaks any notion of using actions as atomic transactions; see [this sandbox](https://codesandbox.io/s/silly-grothendieck-0nyvo).


I'm not sure why you would insist on using async/await _inside_ the action... you can still use it as much as you want outside:
```js
// in the model actions:
fetchAThing: flow(function* fetchAThing(thingId) {
	const thingFromServer = yield fetchThingFromServer(thingId)
	self.thing = thingFromServer
})

// ... somewhere outside the model:
async function fetchThingFromServer(thingId) {
	const response = await callTheServer(thingId)
	const thing = extractThing(response)
	return thing
}
```

**Comment by whinc** [comment]
@beaulac Thank you very much for your answers and examples. Flow is able to keep the action transactional, which is enough to convince me to use flow instead of runInAction + async.

**Comment by jonlambert** [comment]
Also something worth noting is that unlike `await`, the compiler is unable to infer a type from a `yield`. It's not a huge issue but does mean it's not as simple as just swapping one for the other.

**Comment by beaulac** [comment]
Great point – luckily this is [addressed since Typescript 3.6](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-6.html#stricter-generators), so `yield` values always being inferred as `any` is only an issue for TS 3.5 and lower.
